### PR TITLE
Remove a stack frame from .await calls

### PR DESCRIPTION
**Pull request by sfackler**
The stack frames when `.await`ing one async fn from another currently look like this:

```
  12: foo::b::{{closure}}
             at src/main.rs:2
  13: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /home/sfackler/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/future/mod.rs:66
  14: core::future::poll_with_context
             at /home/sfackler/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/future/mod.rs:84
  15: foo::a::{{closure}}
             at src/main.rs:6
```

Since the move away from using TLS to pass the Context around, it's now easy to remove frame 14 by removing poll_with_context in favor of calling Future::poll directly. This still leaves the `GenFuture` frame, but that seems significantly harder to deal with.

It also improves diagnostics a bit since they no longer talk about the private poll_with_context function.

### CODE REVIEW

**Comment by sfackler** [comment]
It looks like the perf run finished but the bot didn't comment for some reason.

**Comment by Centril** [comment]
Yeah, that happens sometimes sadly.

**Review by matthewjasper — COMMENTED** [review]
r=me with nit addressed if we think that the perf is good enough.

File: src/librustc_span/symbol.rs [hunk 1]
```diff
@@ -546,8 +546,9 @@ symbols! {
         plugin,
         plugin_registrar,
         plugins,
+        poll,
         Poll,
-        poll_with_context,
+        get_context,
```
[the remark this answers is not in the document]
Comment (sfackler) [reply src/librustc_span/symbol.rs:551]: Fixed

**Comment by eddyb** [comment]
Post-merge [perf results](https://perf.rust-lang.org/compare.html?start=534a41a32952d36ec73656357777ebbea707aeb4&end=4e4d49d60fd696c4036d438292673a2d7fd34519&stat=instructions:u) are 100% restricted to `await-call-tree`, and it's more clearly >1% now, post-https://github.com/rust-lang/rustc-perf/pull/645 (cc @Mark-Simulacrum).

Presumably we consider that test a stress t est and we're fine with the 1-5% slowdown?
cc @rust-lang/wg-async-foundations

**Pull request merged**